cpp/src/arrow/acero/options_internal.h

```diff
@@ -0,0 +1,37 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#ifndef NDEBUG
+
+#include <mutex>
+#include <ostream>
+
+namespace arrow {
+namespace acero {
+
+struct DebugOptions {
+  DebugOptions(std::ostream* os, std::mutex* mutex) : os(os), mutex(mutex) {}
+
+  std::ostream* os;
+  std::mutex* mutex;
+};
```

Comment:
Instead of passing these all around in options can we just make it part of the query context & query options?

Reply:
I chose this design to allow the caller to limit the debug-mode to a specific execution node. In a future PR, we could add such arguments in the query context/options as a default for all nodes. Let me know what you prefer.

Reply:
If a user needs the ability to toggle debug logging on a per-node basis then I think that would be better done by passing a list of "enabled node ids" to query context / query options.  For example, in most logging frameworks, I can pass in severity level per-category.  So I can, for example, disable Foo logging in some application, but I do that when I'm configuring logging and not when I'm configuring the Foo feature.  Future PR is ok.